File: strands-ts/src/vended-tools/use-agent/use-agent.ts
/**
 * use_agent vended tool: delegate a task to a nested agent.
 *
 * The calling agent constructs a fresh Agent with the given system prompt
 * and an allowlisted subset of the parent's tools, then hands it a single
 * task. The nested agent's final text response is returned to the parent.
 *
 * This is a shim over Agent and its normal construction path; it does
 * not reinvent agent lifecycle. It only enforces the security surface of
 * runtime delegation (allowlist, size caps, recursion cap, cancellation
 * propagation).
 */

import { z } from 'zod'

import { Agent } from '../../agent/agent.js'
import { tool } from '../../tools/tool-factory.js'
import type { InvokableTool } from '../../tools/tool.js'
import type { UseAgentResult, UseAgentStatus } from './types.js'

const DEFAULT_MAX_SYSTEM_PROMPT_BYTES = 8 * 1024
const DEFAULT_MAX_TASK_BYTES = 32 * 1024
const DEFAULT_MAX_TOOL_ALLOWLIST = 64

export const MULTIAGENT_DEPTH_KEY = 'multiagentDepth'
export const MAX_MULTIAGENT_DEPTH = 3

const WILDCARD_TOOL_NAMES: ReadonlySet<string> = new Set(['*', '**', 'all', 'any', ''])

/**
 * Defense-in-depth: child agents cannot invoke multi-agent tools directly. A
 * developer who wants that must register the variants at construction time.
 */
const MULTIAGENT_TOOL_NAMES: ReadonlySet<string> = new Set(['use_agent', 'swarm', 'graph', 'a2a_client'])

const USE_AGENT_DESCRIPTION =
  'Delegate a single task to a nested agent that you construct at call time. ' +
  "You provide the child agent's systemPrompt, an explicit allowlist of tool " +
  'names to expose (drawn from your own tools), and the task itself. The child ' +
  'runs with the same model as you and a fresh conversation, and returns its ' +
  'final text response. Prefer this for scoped sub-tasks that benefit from a ' +
  'different system prompt or a narrower tool surface than your own.'

/**
 * Raised when the shared multi-agent recursion counter has already reached the
 * configured cap. Kept separate from a plain Error so callers that want to
 * catch the depth-cap case can do so without brittle string matching.
 */
export class MultiagentDepthExceededError extends Error {
  constructor(message: string) {
    super(message)
    this.name = 'MultiagentDepthExceededError'
  }
}

function byteLength(value: string): number {
  return new TextEncoder().encode(value).length
}

function validatePositiveIntCap(value: number, name: string): number {
  if (typeof value !== 'number' || !Number.isFinite(value) || !Number.isInteger(value) || value <= 0) {
    throw new Error(`${name} must be a positive integer, got ${value}`)
  }
  return value
}

function validateBoundedString(value: unknown, name: string, maxBytes: number): string {
  if (typeof value !== 'string') {
    throw new Error(`${name} must be a string, got ${typeof value}`)
  }
  if (value.trim().length === 0) {
    throw new Error(`${name} must be non-empty`)
  }
  const size = byteLength(value)
  if (size > maxBytes) {
    throw new Error(`${name} exceeds size cap: ${size} bytes > ${maxBytes} bytes`)
  }
  return value
}

function validateToolAllowlist(tools: unknown, parentToolNames: Set<string>, maxEntries: number): string[] {
  if (tools === undefined || tools === null) return []
  if (!Array.isArray(tools)) {
    throw new Error(`tools must be an array of tool names, got ${typeof tools}`)
  }
  if (tools.length > maxEntries) {
    throw new Error(`tools allowlist exceeds cap of ${maxEntries} entries`)
  }

  const seen = new Set<string>()
  const resolved: string[] = []
  for (const entry of tools) {
    if (typeof entry !== 'string') {
      throw new Error(`tools entries must be strings, got ${typeof entry}`)
    }
    const stripped = entry.trim()
    if (WILDCARD_TOOL_NAMES.has(stripped.toLowerCase())) {
      throw new Error(`tools entry ${JSON.stringify(entry)} is a wildcard; every child tool must be named explicitly`)
    }
    if (MULTIAGENT_TOOL_NAMES.has(stripped)) {
      throw new Error(
        `tools entry ${JSON.stringify(stripped)} is a multi-agent tool and cannot be nested inside a child agent`
      )
    }
    if (seen.has(stripped)) continue
    if (!parentToolNames.has(stripped)) {
      throw new Error(`tools entry ${JSON.stringify(stripped)} is not present in the parent agent's tool registry`)
    }
    seen.add(stripped)
    resolved.push(stripped)
  }
  return resolved
}

function currentDepth(invocationState: Record<string, unknown> | undefined): number {
  const raw = invocationState?.[MULTIAGENT_DEPTH_KEY]
  if (typeof raw !== 'number' || !Number.isFinite(raw) || !Number.isInteger(raw) || raw < 0) return 0
  return raw
}

function mapStopReason(stopReason: string): UseAgentStatus {
  // Shared multi-agent dialect uses the lower-cased SDK Status enum values.
  // Any non-endTurn, non-cancelled, non-interrupt stop reason surfaces as
  // failed so the parent can distinguish a delivered delegation from one
  // that hit a policy or limit.
  if (stopReason === 'endTurn') return 'completed'
  if (stopReason === 'cancelled') return 'cancelled'
  if (stopReason === 'interrupt') return 'interrupted'
  return 'failed'
}

/**
 * Options for {@link makeUseAgent}. All caps default to the shared multi-agent
 * dialect values and are fixed at tool construction; the model cannot alter
 * them at call time.
 */
export interface MakeUseAgentOptions {
  /** Tool name. Defaults to `"use_agent"`. */
  name?: string
  /** Description shown to the model. */
  description?: string
  /**
   * Cap on the shared multi-agent recursion counter. Defaults to
   * {@link MAX_MULTIAGENT_DEPTH}.
   */
  maxDepth?: number
  /** UTF-8 byte cap on `systemPrompt`. */
  maxSystemPromptBytes?: number
  /** UTF-8 byte cap on `task`. */
  maxTaskBytes?: number
  /** Cap on the number of entries in the tool allowlist. */
  maxToolAllowlist?: number
}

/**
 * Build a use_agent vended tool with configurable safety caps.
 *
 * The default caps match the shared multi-agent dialect. They are exposed here
 * so callers can construct the tool with a tighter envelope at wire-up time;
 * the model cannot reach them at call time.
 */
export function makeUseAgent(options: MakeUseAgentOptions = {}): InvokableTool<unknown, UseAgentResult> {
  const {
    name = 'use_agent',
    description = USE_AGENT_DESCRIPTION,
    maxDepth = MAX_MULTIAGENT_DEPTH,
    maxSystemPromptBytes = DEFAULT_MAX_SYSTEM_PROMPT_BYTES,
    maxTaskBytes = DEFAULT_MAX_TASK_BYTES,
    maxToolAllowlist = DEFAULT_MAX_TOOL_ALLOWLIST,
  } = options

  validatePositiveIntCap(maxDepth, 'maxDepth')
  validatePositiveIntCap(maxSystemPromptBytes, 'maxSystemPromptBytes')
  validatePositiveIntCap(maxTaskBytes, 'maxTaskBytes')
  validatePositiveIntCap(maxToolAllowlist, 'maxToolAllowlist')

  const useAgentInputSchema = z
    .object({
      systemPrompt: z
        .string()
        .describe(`System prompt for the nested agent. Non-empty; capped at ${maxSystemPromptBytes} UTF-8 bytes.`),
      task: z.string().describe(`The task to hand the nested agent. Non-empty; capped at ${maxTaskBytes} UTF-8 bytes.`),
      tools: z
        .array(z.string())
        .optional()
        .describe(
          'Exact-name allowlist of tools to expose to the nested agent. Every entry must be a tool that exists ' +
            "in your own tool registry. Wildcards and multi-agent tool names ('use_agent', 'swarm', 'graph', " +
            "'a2a_client') are rejected."
        ),
    })
    .strict()

  return tool({
    name,
    description,
    inputSchema: useAgentInputSchema,
    callback: async (input, context): Promise<UseAgentResult> => {
      if (!context) {
        throw new Error('use_agent requires a tool context')
      }

      const parentAgent = context.agent
      const parentInvocationState = (context.invocationState ?? {}) as Record<string, unknown>

      const depth = currentDepth(parentInvocationState)
      if (depth >= maxDepth) {
        throw new MultiagentDepthExceededError(
          `use_agent refused: recursion depth cap of ${maxDepth} reached (current depth ${depth})`
        )
      }

      const systemPrompt = validateBoundedString(input.systemPrompt, 'systemPrompt', maxSystemPromptBytes)
      const task = validateBoundedString(input.task, 'task', maxTaskBytes)

      const parentToolNames = new Set(parentAgent.toolRegistry.list().map((t) => t.name))
      const toolNames = validateToolAllowlist(input.tools, parentToolNames, maxToolAllowlist)
      const childTools = toolNames.map((toolName) => parentAgent.toolRegistry.get(toolName)!)

      // Child inherits the parent's model instance. Shipped providers are
      // stateless on `this` per invocation, so sharing is safe today; if a
      // future provider retains per-invocation state on the instance, this
      // assumption needs revisiting (e.g. by cloning before delegation).
      //
      // LocalAgent doesn't expose `name` on the shared interface, but the
      // concrete Agent class always sets one (falls back to a default). Read
      // it defensively so this stays compatible with any LocalAgent.
      const parentName = (parentAgent as { name?: string }).name ?? 'agent'
      const child = new Agent({
        model: parentAgent.model,
        systemPrompt,
        tools: childTools,
        name: `${parentName}::use_agent`,
        printer: false,
      })

      // Preserve the parent's invocationState so tracing / telemetry /
      // per-run keys flow through to the child. Only override the shared
      // depth counter with the incremented value.
      const childInvocationState: Record<string, unknown> = {
        ...parentInvocationState,
        [MULTIAGENT_DEPTH_KEY]: depth + 1,
      }

      const cancelSignal = parentAgent.cancelSignal as AbortSignal | undefined

      const start = Date.now()
      const result = await child.invoke(task, {
        invocationState: childInvocationState,
        ...(cancelSignal !== undefined && { cancelSignal }),
      })
      const executionTimeMs = Date.now() - start

      // The child agent surfaces cancellation as stopReason 'cancelled'. Per
      // the shared multi-agent conventions, the TypeScript side re-raises an
      // AbortError so callers can distinguish cancellation from other failures
      // via `error.name === 'AbortError'`, matching the sibling http-request
      // tool. This is a deliberate divergence from Python (which returns
      // `{status: "cancelled"}`); the conventions doc records the asymmetry.
      if (result.stopReason === 'cancelled') {
        const reason = cancelSignal?.reason
        if (reason instanceof Error) {
          throw reason
        }
        throw new DOMException('use_agent cancelled by parent agent', 'AbortError')
      }

      return {
        status: mapStopReason(result.stopReason),
        output: result.toString(),
        executionTimeMs,
      }
    },
  })
}

/**
 * Default use_agent tool with the shared multi-agent dialect's caps. Use
 * {@link makeUseAgent} to construct one with a custom envelope.
 */
export const useAgent = makeUseAgent()

File: strands-py/src/strands/vended_tools/use_agent/use_agent.py
"""Vended ``use_agent`` tool: delegate a task to a nested agent.

The calling agent constructs a fresh :class:`~strands.Agent` with the given
system prompt and an allowlisted subset of the parent's tools, then hands
it a single task. The nested agent's final text response is returned to the
parent.

This is a shim over :class:`~strands.Agent` and its normal construction path;
it does not reinvent agent lifecycle. It only enforces the security surface
of runtime delegation (allowlist, size caps, recursion cap, cancellation
propagation).
"""

from __future__ import annotations

import asyncio
import contextlib
import logging
import math
import time
from typing import TYPE_CHECKING, Any

from ...tools.decorator import tool
from ...types.tools import ToolContext

if TYPE_CHECKING:
    from ...tools.decorator import DecoratedFunctionTool

logger = logging.getLogger(__name__)

_MAX_SYSTEM_PROMPT_BYTES = 8 * 1024
_MAX_TASK_BYTES = 32 * 1024
_MAX_TOOL_ALLOWLIST = 64

_DEPTH_KEY = "multiagent_depth"
_MAX_DEPTH = 3

_WILDCARD_TOOL_NAMES = frozenset({"*", "**", "all", "any", ""})

# Defense-in-depth: child agents cannot invoke multi-agent tools directly. A
# developer who wants that must register the variants at construction time.
_MULTIAGENT_TOOL_NAMES = frozenset({"use_agent", "swarm", "graph", "a2a_client"})


USE_AGENT_DESCRIPTION = (
    "Delegate a single task to a nested agent that you construct at call time. "
    "You provide the child agent's system_prompt, an explicit allowlist of tool "
    "names to expose (drawn from your own tools), and the task itself. The child "
    "runs with the same model as you and a fresh conversation, and returns its "
    "final text response. Prefer this for scoped sub-tasks that benefit from a "
    "different system prompt or a narrower tool surface than your own."
)


class MultiagentDepthExceeded(ValueError):
    """Raised when the shared multi-agent recursion cap is reached."""


def _validate_positive_int_cap(value: Any, name: str) -> int:
    """Validate a factory-time positive integer cap.

    Rejects bool (a Python ``int`` subclass), non-integers, non-finite floats,
    and non-positive values. Keeps the cap-bypass surface tight even when a
    caller wires up the tool with something exotic.
    """
    if isinstance(value, bool):
        raise ValueError(f"{name} must be a positive int, got bool")
    if isinstance(value, float):
        if not math.isfinite(value):
            raise ValueError(f"{name} must be a finite positive int, got {value}")
        if not value.is_integer():
            raise ValueError(f"{name} must be an int, got {value}")
        value = int(value)
    if not isinstance(value, int):
        raise ValueError(f"{name} must be a positive int, got {type(value).__name__}")
    if value <= 0:
        raise ValueError(f"{name} must be positive, got {value}")
    return value


def _validate_bounded_string(value: Any, name: str, max_bytes: int) -> str:
    if not isinstance(value, str):
        raise ValueError(f"{name} must be a string, got {type(value).__name__}")
    if not value.strip():
        raise ValueError(f"{name} must be non-empty")
    size = len(value.encode("utf-8"))
    if size > max_bytes:
        raise ValueError(f"{name} exceeds size cap: {size} bytes > {max_bytes} bytes")
    return value


def _validate_tool_allowlist(tools: Any, parent_tool_names: set[str], max_entries: int) -> list[str]:
    """Validate ``tools`` is an exact-name allowlist drawn from the parent registry.

    Rejects wildcards, inherit-all, non-list values, non-string entries,
    over-long lists, names not in the parent's registry, and multi-agent tool
    names. Deduplicates while preserving order.
    """
    if not isinstance(tools, list):
        raise ValueError(f"tools must be a list of tool names, got {type(tools).__name__}")
    if len(tools) > max_entries:
        raise ValueError(f"tools allowlist exceeds cap of {max_entries} entries")

    seen: set[str] = set()
    resolved: list[str] = []
    for entry in tools:
        if not isinstance(entry, str):
            raise ValueError(f"tools entries must be strings, got {type(entry).__name__}")
        stripped = entry.strip()
        if stripped.lower() in _WILDCARD_TOOL_NAMES:
            raise ValueError(f"tools entry {entry!r} is a wildcard; every child tool must be named explicitly")
        if stripped in _MULTIAGENT_TOOL_NAMES:
            raise ValueError(
                f"tools entry {stripped!r} is a multi-agent tool and cannot be nested inside a child agent"
            )
        if stripped in seen:
            continue
        if stripped not in parent_tool_names:
            raise ValueError(f"tools entry {stripped!r} is not present in the parent agent's tool registry")
        seen.add(stripped)
        resolved.append(stripped)
    return resolved


def _current_depth(invocation_state: dict[str, Any]) -> int:
    """Read the current multi-agent recursion depth from the parent's state.

    ``bool`` is a Python ``int`` subclass; reject it explicitly so ``True``
    doesn't silently count as depth 1. Non-integers and negatives fall through
    to zero: a hostile ``invocation_state`` shouldn't be able to confuse the
    counter into refusing at a lower depth than configured, and a misbehaving
    parent shouldn't crash the tool either.
    """
    raw = invocation_state.get(_DEPTH_KEY, 0)
    if isinstance(raw, bool) or not isinstance(raw, int) or raw < 0:
        return 0
    return int(raw)


def _resolve_parent_tools(agent: Any, names: list[str]) -> list[Any]:
    registry = agent.tool_registry.registry
    return [registry[n] for n in names]


async def _watch_parent_cancel(parent_agent: Any, sub_agent: Any) -> None:
    """Poll the parent's cancel signal and forward it to ``sub_agent``.

    The parent's ``_cancel_signal`` is a ``threading.Event``; sub-agents get
    their own. Cross-agent sharing of the underlying event is not supported,
    so we observe the parent's flag on a short interval and call ``cancel()``
    on the child. Cancel is idempotent.
    """
    while not parent_agent._cancel_signal.is_set():
        await asyncio.sleep(0.05)
    sub_agent.cancel()


def make_use_agent(
    *,
    name: str = "use_agent",
    description: str = USE_AGENT_DESCRIPTION,
    max_depth: int = _MAX_DEPTH,
    max_system_prompt_bytes: int = _MAX_SYSTEM_PROMPT_BYTES,
    max_task_bytes: int = _MAX_TASK_BYTES,
    max_tool_allowlist: int = _MAX_TOOL_ALLOWLIST,
) -> DecoratedFunctionTool:
    """Create the use_agent vended tool.

    Every cap is fixed at the tool boundary and not reachable by the model at
    call time; the parent chooses them at tool construction, exactly like
    ``max_iterations`` on a plain ``Agent``.

    Args:
        name: Tool name. Defaults to ``"use_agent"``.
        description: Description shown to the model.
        max_depth: Cap on the shared multi-agent recursion counter before the
            tool refuses (see ``_multiagent_conventions.md``).
        max_system_prompt_bytes: UTF-8 byte cap on the child's system prompt.
        max_task_bytes: UTF-8 byte cap on the task string.
        max_tool_allowlist: Cap on the number of entries in the tool allowlist.

    Returns:
        A decorated tool that constructs and invokes a nested child agent.
    """
    max_depth = _validate_positive_int_cap(max_depth, "max_depth")
    max_system_prompt_bytes = _validate_positive_int_cap(max_system_prompt_bytes, "max_system_prompt_bytes")
    max_task_bytes = _validate_positive_int_cap(max_task_bytes, "max_task_bytes")
    max_tool_allowlist = _validate_positive_int_cap(max_tool_allowlist, "max_tool_allowlist")

    @tool(name=name, description=description, context="tool_context")
    async def use_agent_tool(
        system_prompt: str,
        task: str,
        tool_context: ToolContext,
        tools: list[str] | None = None,
    ) -> dict[str, Any]:
        """Delegate a single task to a freshly constructed nested agent.

        The nested agent inherits the parent's model instance, runs on a fresh
        conversation, and is exposed only to the parent tools you name in ``tools``.

        Providers shipped with Strands do not hold per-invocation state on
        ``self`` today, so sharing the model instance is safe; if a future
        provider does, this may need to be revisited.

        Args:
            system_prompt: System prompt for the nested agent. Non-empty, capped
                at ``max_system_prompt_bytes`` UTF-8 bytes.
            task: The task to hand the nested agent. Non-empty, capped at
                ``max_task_bytes`` UTF-8 bytes.
            tool_context: Injected by the framework. Not user-facing.
            tools: Exact-name allowlist of tools to expose to the nested agent.
                Every entry must be a tool that exists in your tool registry.
                Wildcards and multi-agent tool names are rejected.
        """
        from ...agent.agent import Agent

        parent_agent = tool_context.agent
        invocation_state = tool_context.invocation_state or {}

        depth = _current_depth(invocation_state)
        if depth >= max_depth:
            raise MultiagentDepthExceeded(
                f"use_agent refused: recursion depth cap of {max_depth} reached (current depth {depth})"
            )

        system_prompt_v = _validate_bounded_string(system_prompt, "system_prompt", max_system_prompt_bytes)
        task_v = _validate_bounded_string(task, "task", max_task_bytes)

        parent_tool_names = set(parent_agent.tool_registry.registry.keys())
        tool_names = _validate_tool_allowlist(tools or [], parent_tool_names, max_tool_allowlist)
        child_tools = _resolve_parent_tools(parent_agent, tool_names)

        sub_agent = Agent(
            model=parent_agent.model,
            system_prompt=system_prompt_v,
            tools=child_tools,
            name=f"{getattr(parent_agent, 'name', 'agent')}::use_agent",
            callback_handler=None,
        )

        # Preserve the parent's invocation_state so tracing / telemetry /
        # per-run keys flow through to the child. Only override the shared
        # depth counter with the incremented value.
        child_invocation_state: dict[str, Any] = {
            **invocation_state,
            _DEPTH_KEY: depth + 1,
        }

        start = time.monotonic()
        watcher = asyncio.create_task(_watch_parent_cancel(parent_agent, sub_agent))
        try:
            result = await sub_agent.invoke_async(task_v, invocation_state=child_invocation_state)
        finally:
            watcher.cancel()
            with contextlib.suppress(asyncio.CancelledError):
                await watcher

        elapsed_ms = int((time.monotonic() - start) * 1000)

        return {
            "status": _map_stop_reason(result.stop_reason),
            "output": str(result),
            "execution_time_ms": elapsed_ms,
        }

    return use_agent_tool


def _map_stop_reason(stop_reason: str) -> str:
    """Map an ``AgentResult.stop_reason`` into the shared multi-agent result vocabulary.

    The shared dialect (see ``_multiagent_conventions.md``) uses the lower-cased
    values of the SDK's multi-agent ``Status`` enum: completed, failed,
    cancelled, interrupted. Any non-``end_turn`` non-``cancelled`` non-``interrupt``
    stop reason (``limit_turns``, ``content_filtered``, ``max_tokens``,
    ``guardrail_intervened``, ...) surfaces as ``failed`` so the parent can
    distinguish a delivered delegation from one that hit a policy or limit.
    """
    if stop_reason == "end_turn":
        return "completed"
    if stop_reason == "cancelled":
        return "cancelled"
    if stop_reason == "interrupt":
        return "interrupted"
    return "failed"


use_agent = make_use_agent()
"""Default use_agent tool with the built-in safety caps."""
